### PR TITLE
Add a function to recover from old versions of saves

### PR DESCRIPTION
**Pull request by Gliese852**
[recording.webm](https://github.com/pioneerspacesim/pioneer/assets/18342621/f38107d6-35ca-45ab-ad8d-8c5049211509)

If the saveload menu decides that it cannot load a save (most likely due to a version mismatch), the `Load` button will be replaced with `Recover` and you can try to restore the game.

In fact, this is the start of a new game, but we are trying to read the parameters from the save. After a recovery attempt, the same dialog for starting a new game opens, but without selecting a start option.

If an error occurs when reading a parameter from a save, or after restoring the parameter is in an invalid state, it is unlocked so that the player can manually specify it. The exception is the flight log - it is always locked.

Version 89 is currently being successfully restored (2023-02-03). For older versions, it is planned to add "reader" versions in the future. Added the ability to create a "versioned" function; when it is launched, dispatching occurs depending on the specified save version.

Also, now when you launch a new game window in debug mode (with ctrl), it immediately switches to "Custom start". In this case, there are no changes to the parameters after the previous launch of the new game window. That is, if you previously ran a recovery, you can open a new game window in debug mode and there will be the same options, but unlocked.

### CODE REVIEW

**Comment by impaktor** [comment]
It would be absolutely fantastic to have this in master before next "stable" release (in less than a month now!). 

My thoughts based on what I saw in the video:

(italics indicate suggested change)
1. could it be "_Minimum_ supported game save version 90" instead of "Actual"?
2. and continue the "unsuccessfully restored parameters are _now_ unlocked, _please set them manually_"

In what way will this collide with @JonBooth78's PR #5666? Just checking so he knows if there's something he should do before we merge that PR.

Great work!

**Comment by Gliese852** [comment]
@impaktor Thanks! For questions:
> 1. could it be "_Minimum_ supported game save version 90" instead of "Actual"?

The fact is that in a sense, all saves are supported, there is no minimum version, it’s just that the older the save, the more errors there will be when restoring. And although the "readers" are written based on version 89, perhaps older versions will also be restored somehow. I have a save of version 86 (October 11, 2020), "only" cargo and equipment failed to be restored from it.
> 2. and continue the "unsuccessfully restored parameters are _now_ unlocked, _please set them manually_"

placed in TODO, there might be more similar comments

**Comment by impaktor** [comment]
> The fact is that in a sense, all saves are supported, there is no minimum version

I guess this comes down to semantics in the end. Either way, I think calling version 90 "Current version" or "Minimum version" is preferable over "Actual version", as I don't think the use of the word "Actual" is optimal here.

**Comment by JonBooth78** [comment]
I don't think this will make much difference to my change in #5666 - if it does I can account for that as I'll need to rebase it anyway now as things have moved on.

**Comment by JonBooth78** [comment]
Hmm, OK, I'll need to look at that then; since the new version was written to read with backwards compatibility I'd hoped it would 'just work'...

Moving forwards then, after this change to give us backward compatibility on old save versions will we need to write multiple code paths?

**Comment by Web-eWorks** [comment]
Note: we will be savebumping to v90 as part of this release. I'd ideally like to merge this PR "last" after all other major feature PRs to keep the amount of rewriting required to a minimum.

**Comment by Gliese852** [comment]
@JonBooth78 Yes, I saw you provided backwards compatibility, but unfortunately, backwards compatibility won't work because as soon as the save version is increased, the save simply won't load, at least that's how it works now. 
It would probably be possible to change this by maintaining backwards compatibility at all times, ensuring that the previous version of a save will load in the next one, but it seems that maintaining compatibility through a recovery system should be less expensive since a significant amount of data is ignored.

**Comment by Gliese852** [comment]
@JonBooth78 
> Moving forwards then, after this change to give us backward compatibility on old save versions will we need to write multiple code paths?

If I understood the question correctly, "inside the game code", you do not need to maintain any backward compatibility, and your code can be quite straightforward.

But it is required to maintain the correct operation of changed "starting parameter" in the start menu, including editing in the "Custom Game" mode and restoring from a save.

Here is a list of starting parameters to make it more clear what I’m talking about:
https://github.com/Gliese852/pioneer/blob/1ccd9a20e635924474b3831c5266bd49841ceef3/data/pigui/modules/new-game-window/layout.lua#L15

**Comment by Gliese852** [comment]
I thought it would be useful to be able to also restore saves from the current version, for testing purposes or if an incompatible change occurs between releases. Therefore, I added this thing - if you hold Ctrl, the button forcibly changes to "Recover".

**Review by impaktor — COMMENTED** [review]
Just went through the strings.

File: data/lang/ui-core/en.json
Comment (Web-eWorks) [reply data/lang/ui-core/en.json:2200]: I'd recommend `System body index is out of range` instead - because that's the "real" problem. It's an invalid index, not a too-large index.

**Comment by Gliese852** [comment]
Rebased and removed conflicts.

**Comment by JonBooth78** [comment]
I now see what you've done here - very nice.

I was playing around with this trying to load some old save games that were made unloadable during this dev cycle.

Turns out the flight log for the stations isn't getting loaded properly with graceful degradation as at least one of the stations appears to be on an invalid index planet.

(I was looking to see if I could help integrate this with the updated flightlog code...)

Attached:

[jump to pirates.zip](https://github.com/pioneerspacesim/pioneer/files/14000597/jump.to.pirates.zip)

**Comment by Gliese852** [comment]
@JonBooth78 Thank you! I'll do something about it.

**Comment by Gliese852** [comment]
Fixed the problem, but only in the recovery logic and the start menu - inside the game it will crash when going to the station log, since it does not know how to show "outdated" system paths. Haven’t touched the GUI in the game yet to avoid unnecessary conflicts with #5666.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/recovery.lua:17]: This should probably be made public, as I'd want to write savefile migration functions in their own file which `require()s` the `Recovery` module.

File: data/pigui/modules/new-game-window/ship.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/ship.lua:840]: Note: with EquipSet v2 (hopefully this doesn't offend @laarmen - I believe there was only ever 1 EquipSet implementation...), equipment will be migrated to a new "persistent object" feature stored in `lua_modules_persistent`. (It will also need a new frontend for setting equipment when starting the game, but that's a different topic entirely).

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/recovery.lua:24]: Note: sooner or later, this function will be required to post-process the save game and fix up all contained table references. The order and location a table value *definition* is serialized in is non-deterministic and any JSON value encoding a reference to a specific table value may potentially be the one in which that table is serialized to.

As a result, I'd recommend this function go through all `{ ref: number }` values serialized in the `lua_modules_json` (and eventually `lua_modules_persistent`) tables and perform a partial unserialization / unpickling, enough to ensure the savefile is traversable as a "normal" Lua table with the appropriate table identities preserved.

File: src/lua/LuaJson.cpp
Comment (Web-eWorks) [inline src/lua/LuaJson.cpp:42]: Thanks for catching this - I hadn't actually exercised this codepath until writing the equipment rework.

File: src/lua/LuaSystemBody.cpp
Comment (Web-eWorks) [inline src/lua/LuaSystemBody.cpp:155]: Note: `GetStarSystem` does not maintain a strong reference to the returned StarSystem (just a weak one so it can be removed from the master cache when all other references expire), so this system (AFAIK) will be deleted as soon as the `if` block scope is closed. This would orphan the just-created system body, leaving it still without parent.

If this is behaving differently than I've described, something's holding on to a reference to the star system that's been created and I retract my comment.

File: src/savegamedump.cpp
Comment (Web-eWorks) [inline src/savegamedump.cpp:29]: Note: I prefer to run dumped savegames through https://jsonviewer.stack.hu/, but this is a good secondary option if your text editor is happy with highlighting 1MB+ JSON files.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]
I'm noticing a general pattern that `reader` functions do two things: they define a path in the savefile to unpickle data from, and then they convert the unpickled saved data into a "live" object.

I wonder if it'd be worthwhile to evolve this pattern to treat those as two separate processes, such that there is one function concerned only with loading unpickled data from the current save version, and there is another (versioned) function concerned with finding the data from the save game and unpickling it into the format currently expected by the game.

In a lot of cases, readers seem to boil down to a simple "passthrough" function which loads data from a (set of) specific key(s), unpickles it, and optionally transforms it before assigning to the parameter value. Some of this functionality could be made implicit, I think.

This is probably not a topic for this iteration of the save game restoration however, but I did want to record the idea for potential future improvements.

-----

For more on-topic discourse, I've left several review comments and change requests where there are things that I'd like to see addressed before merge. As this PR has a (somewhat) low bug surface and extremely high value for the upcoming release I'm willing to merge it even after we enter feature freeze if they can't be addressed before then, but "the sooner the better" is the motto of the remaining 8 days before release. :smile:

Apologies that this has taken so long to review - I've not had mental bandwidth available to deal with the review process for most of this week. In general, I have high confidence in the quality of code being written here, and since it's mostly touching code you've written yourself I don't have to worry about it not matching the existing style of (this particular corner of) the codebase. :smile:

**Comment by Gliese852** [comment]
Made a rebase to resolve the conflict, minimally adapted to the new flightlog, it works, but there are 2 problems. Firstly, the entries in the game are grouped by type in violation of chronology. Secondly, the log inside the game cannot show a "non-existent" system/station path, an error occurs.

**Comment by Gliese852** [comment]
Fixed the above 2 problems and it is now everything is working.

File: src/lua/LuaSystemBody.cpp
Comment (Gliese852) [reply src/lua/LuaSystemBody.cpp:155]: Yes, you're right! Probably nothing broke for me, because in fact I never managed to get into this branch when testing the game. Apparently, it is enough for the system to survive until the end of the function.

File: src/savegamedump.cpp
Comment (Gliese852) [reply src/savegamedump.cpp:29]: Fortunately, I don't experience any problems with large files (in particular json) in vim. And all my bindings with me!

File: data/pigui/libs/message-box.lua
Comment (Gliese852) [reply data/pigui/libs/message-box.lua]: Done!

File: data/pigui/modules/new-game-window/class.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/class.lua]: Done!

File: data/pigui/modules/new-game-window/crew.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/crew.lua]: Done!

File: data/pigui/modules/new-game-window/crew.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/crew.lua]: Done!

File: data/pigui/modules/new-game-window/crew.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/crew.lua:485]: Done!

File: data/pigui/modules/new-game-window/helpers.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/helpers.lua]: Done!

File: data/pigui/modules/new-game-window/helpers.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/helpers.lua]: Done!

File: data/pigui/modules/new-game-window/location.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/location.lua:321]: Inserted blank lines, yes, that’s better.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua]: Done!

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua:83]: I did the substitution of refs, but did not do the unpickling - all readers unpickle what they need themselves, is it worth redoing? It seems that the reader code will not be simplified by this.
Although, after resolving the refs, I can simplify the equipment reader - there was a search using refs, which is now not needed.

File: data/pigui/modules/new-game-window/location.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/location.lua:342]: Unless you're using this `paths` variable somewhere else, I'd recommend using the attr `starSystem.numberOfBodies` instead - it doesn't push new SystemPath objects to Lua for every (potentially 70+) body in the system.

File: data/pigui/modules/new-game-window/location.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/location.lua:340]: This seems *very* prone to breakage and dependent on internal logic. At the least, I think this should have a longer documentation comment indicating the purpose of the `dockedTo` variable here and *why* it's needed to subtract it to match or not match the values contained in a private array member in C++.

**Review by Web-eWorks — CHANGES_REQUESTED** [review]

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json:1896]: Oh, it turns out this line is no longer needed!

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Done!

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Named as 'Current'.

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Done!

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Done!

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Added e.g. in the description

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Done!

File: data/lang/ui-core/en.json
Comment (Gliese852) [reply data/lang/ui-core/en.json]: Done!

File: data/pigui/modules/new-game-window/location.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/location.lua]: Done!

File: data/pigui/modules/new-game-window/location.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/location.lua:347]: Unfortunately, I didn't find another way. In my defense, it seems to me that the entire recovery system depends heavily on internal logic. And if something goes wrong, you will at least find yourself in the same system. Also added an explanatory comment.

Comment (Web-eWorks) [reply data/pigui/modules/new-game-window/location.lua:347]: That comment is quite satisfactory, thank you!

File: src/lua/LuaSystemBody.cpp
Comment (Web-eWorks) [reply src/lua/LuaSystemBody.cpp:155]: In general I think the code pattern of interacting with orphaned SystemBody objects should be strongly discouraged, due to the steep cost of re-generating the system on every access to the `parent` member of both the current orphaned body and the returned parent body. However, this is an acceptable solution to prevent total breakage.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua:97]: Did a complete unpickling before calling the readers. Now only the fixupDoc function in the Recovery module is responsible for compatibility with the Lua serializer.

**Comment by Gliese852** [comment]
I think I finished everything I wanted.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/recovery.lua:61]: Note: serialized keys can also be table instances. The in-progress EquipSet v2 came close to using the pattern of serializing `{ [table] = true }` but eventually was refactored away. I'm unsure if we have other occurrences of that pattern.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [inline data/pigui/modules/new-game-window/recovery.lua:37]: This function could handle references as well:

```lua
local k, v = tbl[i], tbl[i + 1]
if type(k) == "table" and k.ref then k = cache[k.ref] end
if type(v) == "table" and v.ref then v = cache[v.ref] end
result[k] = v
```

It could even be handled another way to combine the `getRefs`, `putRefs`, and `unpickle` functions together:

```lua
local cache = {}

local function unpickle(v)
	if type(v) ~= 'table' or not v.ref then return v end
	if not v.table then return cache[v.ref] end

	local result = {}
	cache[v.ref] = result

	for i = 1, #v.table, 2 do
		local key = unpickle(v.table[i])
		local val = unpickle(v.table[i + 1])
		result[key] = val
	end

	return result
end
```

**Review by Web-eWorks — CHANGES_REQUESTED** [review]

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [reply data/pigui/modules/new-game-window/recovery.lua:97]: Creating readers is now less complicated and the pickling mechanism is now just an implementation detail - thanks for tackling that!

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua:61]: Indeed, there are such places, although not in the restored parameters.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua:37]: The proposed version of the function will definitely not work, because the function will end on the first line - `gameDoc` does not have a `ref` field.
It is also assumed that the definition of a lua object will necessarily be encountered before other references to it, but this does not seem to be the case, at least because we are dealing with general Lua table at top level.

Comment (Web-eWorks) [reply data/pigui/modules/new-game-window/recovery.lua:37]: It was meant to illustrate the concept more than be tested/proven-working code. That being said, I see your point about iteration order; although pickled Lua tables internally require that the first encountered instance of a table value contains its definition when unserializing, that may not hold true when operating on recovered save data that's not going through the C++ serialization process.

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Web-eWorks) [reply data/pigui/modules/new-game-window/recovery.lua:61]: Not in the recovered parameters *right now* at least. This may change at any time.

Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua]: Now I take this into account!

File: data/pigui/modules/new-game-window/recovery.lua
Comment (Gliese852) [reply data/pigui/modules/new-game-window/recovery.lua:37]: I picked up some things from your code, but I still have to do two passes.

**Comment by Gliese852** [comment]
This is a clean copy.

**Comment by laarmen** [comment]
Finally, peace and quiet in my inbox!

Nice work folks, that's a nifty feature 🙂

**Pull request merged**
